### ISSUE TITLE
Matrice drifting on VirtualStick FlightControlData with wind

### ISSUE THREAD

**Issue by rabbishuki** [dji-sdk/Mobile-SDK-Android#451]
`DJI Android SDK version: 4.11`
`Matrice 100` / `Matrice 600`

I am trying to take off the drone vertically.

I tried with GoToAction in a timeline, but that failed due to some bug in the SDK, (confirmed by your support team dev@dji.com #29496) I get STARTED for the GoToAction, but no PROGRESSED or FINISHED, and no errors logged at all.

Since I need to continue working, I tried a workaround by sending FlightControlData to the VirtualStick by calling the following function with the requested height 20 times a second:

```
VerticalControlMode.POSITION
FlightOrientationMode.AIRCRAFT_HEADING
VirtualStickModeEnabled = true
VirtualStickAdvancedModeEnabled = true

void sendHeightCommand(Float requestedAltitude) {
    FlightControlData data = new FlightControlData(0f, 0f, 0f, requestedAltitude);
    flightController.sendVirtualStickFlightControlData(data, djiError -> {
        log.v(djiError.getDescription);
    });
}
```

And it works (with the right amount of timeouts) but if there is wind, the drone drifts away, which is very dangerous for me as there is more than one drone in the field, and I don't want them to collide.

Is there another way to change the altitude of the drone, while maintaining its position?
Or is there a way to measure the wind, and push back against it?

Thank you very much

**Comment by kenargo** [comment]
I wrote my own takeoff timeline,  its easy to create you're own. 

If you don't need timeline,  mission will takeoff correctly. 

⁣Get BlueMail for Android ​

**Comment by rabbishuki** [comment]
Can you share an example?

**Comment by kenargo** [comment]
Here you go; you will need to remove calls not in your app but it will be clear I think:

`public class TakeoffTimelineElement extends TimelineElement {

    private final boolean failOnError;

    public TakeoffTimelineElement(final boolean failOnError) {
        this.failOnError = failOnError;
    }

    @Override
    public void run() {

        MyApplication.getLogger().info("TakeoffTimelineElement.run");

        MissionControl.getInstance().onStart(this);

        // If motors are already
        if (!(boolean) KeyManager.getInstance().getValue(FlightControllerKey.create(FlightControllerKey.ARE_MOTOR_ON))) {
            MissionControl.getInstance().onFinishWithError(TakeoffTimelineElement.this, null);
        } else {

            AircraftConnectionViewModel.getAircraftInstance().getFlightController().startTakeoff(djiError -> {
                if (djiError == null) {
                    MissionControl.getInstance().onFinishWithError(TakeoffTimelineElement.this, null);
                } else {
                    MyApplication.getLogger().info("startTakeoff: {}", djiError);
                    MissionControl.getInstance().onFinishWithError(TakeoffTimelineElement.this, djiError);
                }
            });
        }
    }

    @Override
    public boolean isPausable() {
        MyApplication.getLogger().info("TakeoffTimelineElement.isPausable");
        return false;
    }

    @Override
    public void stop() {
        MyApplication.getLogger().info("TakeoffTimelineElement.stop");

        AircraftConnectionViewModel.getAircraftInstance().getFlightController().cancelTakeoff(djiError -> {
            if (djiError == null) {
                MissionControl.getInstance().onFinishWithError(TakeoffTimelineElement.this, null);
            } else {
                MyApplication.getLogger().info("cancelTakeoff: {}", djiError);
                if (failOnError) {
                    MissionControl.getInstance().onFinishWithError(TakeoffTimelineElement.this, djiError);
                } else {
                    MissionControl.getInstance().onFinishWithError(TakeoffTimelineElement.this, null);
                }
            }
        });
    }

    @Override
    public DJIError checkValidity() {
        MyApplication.getLogger().info("TakeoffTimelineElement.checkValidity");
        return null;
    }

    @Override
    public void finishRun(@Nullable DJIError djiError) {
        MyApplication.getLogger().info("TakeoffTimelineElement.finishRun");
    }
}`

**Comment by rabbishuki** [comment]
@kenargo Thanks for your time and effort, it is very appreciated.

I guess I wasn't clear enough so I will explain.

The takeoff part works, but after takeoff, the drone hovers at 1.2 Meters.

The problem is that now I want it to go to 40 Meters high, at the same GPS location, but the `GoToAction` element isn't working.

That is why I tried by using `VirtualStick` but then I have a problem with the wind.

What other options do I have to change the drone's altitude?

**Comment by dji-dev** [comment]
Public comment from Luce Luo in <a href='https://djisdksupport.zendesk.com/agent/tickets/29776'>Zendesk ticket #29776</a>:<br><br>Dear Customer,

Thank you for contacting DJI.
Ticket #29496 has been escalated to the engineer, and we will contact you again once any feedback.
I try to find some APIs for you, please check if any help.
setVirtualStickAdvancedModeEnabled
 ![](https://djisdksupport.zendesk.com/attachments/token/dkk2c41X58TRA8F0y84sKRcti/?name=inline301439834.png)​
getFlightWindWarning
 ![](https://djisdksupport.zendesk.com/attachments/token/HZPMVkXfbjf6AzPGWSwL0ETWP/?name=inline-846939285.png)​


Thanks,

DJI Dev Team<br><a href='https://djisdksupport.zendesk.com/attachments/token/dkk2c41X58TRA8F0y84sKRcti' target='zd_attach'>inline301439834.png</a><br><a href='https://djisdksupport.zendesk.com/attachments/token/HZPMVkXfbjf6AzPGWSwL0ETWP' target='zd_attach'>inline-846939285.png</a>

**Comment by kenargo** [comment]
What I would suggest it creating a waypoint timeline to fly 40m above the drone's location. 

Virtual joystick is a real pain to use,  especially for what you're trying to do. 

⁣Get BlueMail for Android ​

**Comment by rabbishuki** [comment]
@dji-dev Thanks for your response.

I tried `setVirtualStickAdvancedModeEnabled` but didn't help.

as for the `getFlightWindWarning` that will only give me a warning, but I want to know how much wind in order to send a command to the `virtual stick` to stay in the same place.

**Comment by rabbishuki** [comment]
> What I would suggest it creating a waypoint timeline to fly 40m above the drone's location. Virtual joystick is a real pain to use,  especially for what you're trying to do. ⁣Get BlueMail for Android ​

@kenargo A waypoint can be part of the timeline?

**Comment by kenargo** [comment]
FYI, if you're really intent on using virtual joystick,  look at setVirtualStickAdvancedModeEnable.  It's false by default but will allow the drone to stabilize in winds. 

⁣Get BlueMail for Android ​

**Comment by rabbishuki** [comment]
@kenargo I tried `setVirtualStickAdvancedModeEnable` but don't seem to see a difference in the simulator. maybe I put too much wind?

**Comment by rabbishuki** [comment]
@kenargo so after testing in the simulator it seems that `setVirtualStickAdvancedModeEnable(true)` does the job.

I will go out to test the code in the field tomorrow and confirm that it actually works.

Still wish that @dji-dev can get `GoToAction` to work, it seems more fit for the job.

**Comment by kenargo** [comment]
Yes,  in fact having waypoint missions im timelines was a key driver for the timeline functionality. 

The sample actual shows how to include a mission in a timeline. 

⁣Get BlueMail for Android ​

**Comment by kenargo** [comment]
Using the code I sent for takeoff (there is a bug in DJI's version) you could create a small mission that gives you the functionality.

I suspect inside the GoToAction most likely uses a waypoint under the covers. 

⁣Get BlueMail for Android ​
